upgrade.sh

```diff
@@ -7,51 +7,102 @@
+                        --wait --install --force coder coder/coder \
+                        --values tmp/current-values.yml
+                    else
+                        helm upgrade --namespace $NAMESPACE --atomic \
+                        --wait --install --force --version $VERSION \
+                        coder coder/coder --values tmp/current-values.yml   
+                    fi
+                }
+
+                if fix; then
+                    echo "\n\nUpgrade successful."
+                else
+                    echo "\n\nUpgrade failed. Please reference our documentation for troubleshooting
+                    a failed upgrade https://coder.com/docs/setup/updating#fixing-a-failed-upgrade. 
+                    If you still run into issues, contact support@coder.com"
+                fi
+
+                exit >&2         
+        esac
+    done
+    # if flag exists, then do not run the below.
+    else {
+        function prerequisites() {
+        # check if Coder helm repo is added, if not, add coder helm repo
+            if helm repo list | grep -q "https://helm.coder.com"; then
+                echo "Coder already exists within the current helm config."
+            else
+                helm repo add coder https://helm.coder.com
+            fi
+                }
+
+        function upgrade() {
+        # retrieve latest repository information
+            helm repo update
+
+        # upgrade to desired version - if VERSION doesn't exist, 
+        # run upgrade command with VERSION omitted (default to latest).
+            if [ "$VERSION" == "" ]; then
+                helm upgrade --namespace $NAMESPACE --force --install --atomic --wait \
+                coder coder/coder
+            else
+                helm upgrade --namespace $NAMESPACE --force --install --atomic --wait \
+                --version $VERSION coder coder/coder
+            fi
+                }
+
+            function usage() {
+                echo "Usage: set NAMESPACE and VERSION variables before running. 
+                    VERSION format should follow x.xx.x. Omitting VERSION will default to latest."
+                exit 1
+                }
+
+            # return usage if NAMESPACE is blank or VERSION doesn't match the correct format.
+                if [[ "$VERSION" != "" && ! "$VERSION" =~ ^([0-9]\.([1-9]|[1-9][0-9])\.[0-9])$ || "$NAMESPACE" == "" ]]; then
+                    usage
+                fi
+
+            function main() {
+                prerequisites
+
+                if upgrade; then
+                    echo "\n\nUpgrade successful."
+                else
+                    echo "\n\nUpgrade failed. Please reference our documentation for troubleshooting
+                    a failed upgrade https://coder.com/docs/setup/updating#fixing-a-failed-upgrade. 
+                    Alternatively, run ./upgrade.sh --fix to attempt the commands automatically."
```

Comment:
this is awesome! i love how you did everything in functions. makes it very easy to execute and check